Lib/typing.py

```diff
@@ -1387,23 +1381,20 @@ def _no_init(self, *args, **kwargs):
     if type(self)._is_protocol:
         raise TypeError('Protocols cannot be instantiated')
 
-def _callee(depth=2, default=None):
+def _callee(depth=1, default='__main__', unsupported=None):
     try:
-        return sys._getframe(depth).f_globals['__name__']
+        return sys._getframe(depth + 1).f_globals.get('__name__', default)
     except (AttributeError, ValueError):  # For platforms without _getframe()
-        return default
+        return unsupported
 
 
-def _allow_reckless_class_checks(depth=3):
+def _allow_reckless_class_checks(depth=3, sentinel=object()):
     """Allow instance and class checks for special stdlib modules.
 
     The abc and functools modules indiscriminately call isinstance() and
     issubclass() on the whole MRO of a user class, which may contain protocols.
     """
-    try:
-        return sys._getframe(depth).f_globals['__name__'] in ['abc', 'functools']
-    except (AttributeError, ValueError):  # For platforms without _getframe().
-        return True
+    return _callee(depth, unsupported=sentinel) in {'abc', 'functools', sentinel}
```

Comment:
IMO this is less readable than the previous version since usage of `unsupported` is cryptic here and it forced a change to `_callee` which made it more complex.

Reply:
Agree, I have added `unsupported` specially for case like in `_allow_reckless_class_checks`. Should we rename this parameter or return `None` on case when `_getframe` is unsupported?